### PR TITLE
Add Tags command and support finding by tags

### LINKED ISSUE #123: Let users find modules by tags
As a student, I want to be able to search my `Completed Modules` by tags, so that I can use tags to easily organize my modules.

### LINKED ISSUE #124: Let users view all tags
As a student, I want to be able to view all the tags that are present, so that I don't have to scroll through `Completed Modules` manually and see.

### PR DESCRIPTION
**Pull request by Silvernitro**
#### Feature: Let users list out all tags (closes #124)
- Add `TagsCommand`
- Refactor GradPad to use a `UniqueTagMap` to keep track of tags. This allows us to get a list of unique tags without having to loop through every module and checking. This is also what was suggested in the AB3 DG:

Screenshot 2020-10-27 at 3 48 39 PM

- Add a moduleCount field to `Tag`, which keeps track of how many modules contain it. See below for why.
- When a module is added:
  1. See if any of the tags to be added already exist in `UniqueTagMap` and use the existing ones instead.
  2. Else, add all tags that are actually new to `UniqueTagMap`.
- When a module is deleted:
  1. Decrease the moduleCount of all tags in that module.
  2. Only delete a tag from `UniqueTagMap` if its moduleCount is 0

#### Feature: Let `find` command search by tags (closes #123)
- Add `ModuleContainsTagsPredicate`
- Add `CompoundFindPredicate`, which basically combines `ModuleCodeContainsKeywordsPredicate` and `ModuleContainsTagsPredicate`
- Tweak `FindCommandParser` to parse and create a `CompoundFindPredicate` instead
- Tweak `FindCommand` to accept `Predicate<Module>` (actually idk why this wasn't there in the first place to support OCP)

#### Tests:
- Add tests for all new classes written
- Tweak/add tests for existing classes that were tweaked

### CODE REVIEW

**Review by shaokiat — APPROVED** [review]
Wow super impressed with the clean code! 💯 🥇 Nice one Sam!

Can I ask why did you use `Predicate<Module>` to replace `ModuleCodeContainsKeywordsPredicate`? Is this because Predicate<Module> has a broader scope so that you can combine both `ModuleCodeContainsKeywordsPredicate` and `ModuleContainsTagPredicate`?

File: src/main/java/seedu/address/model/module/ModuleContainsTagsPredicate.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+     */
+    public ModuleContainsTagsPredicate(List<String> tagNames) {
+        this.tagNames = tagNames;
+    }
+
+    @Override
+    public boolean test(Module module) {
+        return tagNames.stream().anyMatch(name ->
+              module.getTags().stream().anyMatch(tag ->
+                     tag.tagName.equals(name)));
```
Comment (mhdsyfq77) [inline src/main/java/seedu/address/model/module/ModuleContainsTagsPredicate.java]: Perhaps can improve this by using equalsIgnoreCase? cause as of now if i type in `find foundation`, all the modules with 'Foundation' as tags won't be listed  😢 
```suggestion
                     tag.tagName.equalsIgnoreCase(name)));
```

**Review by mhdsyfq77 — COMMENTED** [review]
everything looks good sam! impressive as always 👏🏽  just one smol suggestion!

File: src/main/java/seedu/address/model/module/ModuleContainsTagsPredicate.java (hunk 1 above)
Comment (Silvernitro) [reply src/main/java/seedu/address/model/module/ModuleContainsTagsPredicate.java]: Very very good catch! Thanks and resolved :D

**Pull request merged**